delphi-frontend/src/test/java/au/com/integradev/delphi/type/intrinsic/IntrinsicArgumentMatcherTest.java

```diff
@@ -18,55 +18,126 @@
-                FACTORY.array("TFoo", TypeFactory.untypedType(), Set.of(ArrayOption.DYNAMIC))))
+  void testLikeDynamicArray() {
+    Type dynamicArray =
+        FACTORY.array("TFoo", TypeFactory.untypedType(), Set.of(ArrayOption.DYNAMIC));
+    Type fixedArray = FACTORY.array("TBar", TypeFactory.untypedType(), Set.of(ArrayOption.FIXED));
+    Type arrayConstructor =
+        FACTORY.arrayConstructor(List.of(FACTORY.getIntrinsic(IntrinsicType.INTEGER)));
+
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, dynamicArray)).isTrue();
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, arrayConstructor)).isTrue();
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, FACTORY.getIntrinsic(IntrinsicType.ANSICHAR))).isTrue();
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, FACTORY.getIntrinsic(IntrinsicType.WIDECHAR))).isTrue();
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, FACTORY.getIntrinsic(IntrinsicType.SHORTSTRING)))
         .isTrue();
-    assertThat(
-            matches(
-                ANY_DYNAMIC_ARRAY,
-                FACTORY.array("TBar", TypeFactory.untypedType(), Set.of(ArrayOption.FIXED))))
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, FACTORY.getIntrinsic(IntrinsicType.ANSISTRING)))
+        .isTrue();
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, FACTORY.getIntrinsic(IntrinsicType.UNICODESTRING)))
+        .isTrue();
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, fixedArray)).isFalse();
+    assertThat(matches(LIKE_DYNAMIC_ARRAY, FACTORY.emptySet())).isFalse();
+  }
+
+  @Test
+  void testAnyString() {
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.ANSICHAR))).isTrue();
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.WIDECHAR))).isTrue();
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.SHORTSTRING))).isTrue();
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.ANSISTRING))).isTrue();
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.UNICODESTRING))).isTrue();
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.PANSICHAR))).isFalse();
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.PCHAR))).isFalse();
+    assertThat(matches(ANY_STRING, FACTORY.getIntrinsic(IntrinsicType.INTEGER))).isFalse();
+  }
+
+  @Test
+  void testAnyFile() {
+    assertThat(matches(ANY_FILE, FACTORY.untypedFile())).isTrue();
+    assertThat(matches(ANY_FILE, FACTORY.getIntrinsic(IntrinsicType.TEXT))).isTrue();
+    assertThat(matches(ANY_FILE, FACTORY.fileOf(FACTORY.getIntrinsic(IntrinsicType.INTEGER))))
+        .isTrue();
+    assertThat(matches(ANY_FILE, FACTORY.getIntrinsic(IntrinsicType.INTEGER))).isFalse();
+  }
+
+  @Test
+  void testAnyTextFile() {
+    assertThat(matches(ANY_TEXT_FILE, FACTORY.untypedFile())).isFalse();
+    assertThat(matches(ANY_TEXT_FILE, FACTORY.getIntrinsic(IntrinsicType.TEXT))).isTrue();
+    assertThat(matches(ANY_TEXT_FILE, FACTORY.fileOf(FACTORY.getIntrinsic(IntrinsicType.INTEGER))))
```

Comment:
I assume we've confirmed that e.g. `file of string` doesn't get counted as a text file by Delphi?

Reply:
`file of string` is not allowed, actually.
`E2155 Type 'string' needs finalization - not allowed in file type`

`Text` is really a sort of weird aberration, it's different from the other file types in a few ways.
It doesn't even have the same size as a normal `file of X` type.